### PR TITLE
Improve detection of resource attributes on ECS

### PR DESCRIPTION
**Pull request by felixscheinost**
This improves the detection of resource attributes on ECS by fetching ECS metadata from `ECS_CONTAINER_METADATA_URI` or `ECS_CONTAINER_METADATA_URI_V4`.

Previously only `CONTAINER_NAME` and `CONTAINER_ID` id were set.

Now we set:

- CONTAINER_ID
- CONTAINER_NAME
- AWS_ECS_CONTAINER_ARN
- CONTAINER_IMAGE_NAME
- CONTAINER_IMAGE_TAG
- aws.ecs.container.image.id
- AWS_LOG_GROUP_ARNS
- AWS_LOG_GROUP_NAMES
- AWS_LOG_STREAM_NAMES
- AWS_ECS_TASK_ARN
- AWS_ECS_TASK_FAMILY
- AWS_ECS_TASK_REVISION

Especially `AWS_LOG_GROUP_ARNS` is important so that connection of traces to logs works OOTB on X-Ray.

### CODE REVIEW

File: sdk-extensions/aws/src/main/java/io/opentelemetry/sdk/extension/aws/resource/EcsResource.java [hunk 1]
```diff
@@ -36,38 +42,182 @@ public static Resource get() {
...
+      // May put the same attribute twice but that shouldn't matter.
+      parseUrl(httpClient, ecsMetadataUrl + "/task", attrBuilders);
+      return Resource.create(attrBuilders.build(), ResourceAttributes.SCHEMA_URL);
     }
+    // Not running on ECS
+    return Resource.empty();
+  }
 
-    AttributesBuilder attrBuilders = Attributes.builder();
+  static void parseUrl(SimpleHttpClient httpClient, String url, AttributesBuilder attrBuilders) {
```
Comment (anuraaga) [inline sdk-extensions/aws/src/main/java/io/opentelemetry/sdk/extension/aws/resource/EcsResource.java:65]:
```suggestion
  static void fetchMetadata(SimpleHttpClient httpClient, String url, AttributesBuilder attrBuilders) {
```

**Review by anuraaga — COMMENTED** [review]
Thanks! This generally looks good, @willarmiros can you also give it a check?

**Comment by willarmiros** [comment]
Thanks so much for this improvement! I will have someone from AWS take a look at this shortly

**Comment by felixscheinost** [comment]
Bump: Please don't close this PR automatically.

**Comment by jkwatson** [comment]
> Thanks so much for this improvement! I will have someone from AWS take a look at this shortly

@willarmiros any update here?

**Comment by willarmiros** [comment]
Apologies for the delay here, we will be having someone review this shortly!

**Comment by mmanciop** [comment]
@felixscheinost i missed out on this PR and worked on #4670. Out two PRs have a lot of overlap, but also some differences in which attributes we set. @willarmiros is going to have a look at these PRs in the next few days. Would you be interested in pulling into your PR (which came first :-) ) the attributes I additionally implemented?

**Comment by felixscheinost** [comment]
@mmanciop Thanks for rebasing your changes on my PR! 

Regarding the change in the log group ARN. That dosn't seem right to me. Have you tested this on AWS?

I just changed this line on my branch, to test this change in isolation, and deployed to our test evironment.

```
      // from
      attributesBuilder.put(ResourceAttributes.AWS_LOG_GROUP_ARNS, "arn:aws:logs:" + region + ":" + account + ":log-group:" + logGroupName);
      // to
      attributesBuilder.put(ResourceAttributes.AWS_LOG_GROUP_ARNS, "arn:aws:logs:" + region + ":" + account + ":log-group:/aws" + logGroupName + ":*");
```

With this change in place I can no longer see the logs for a trace in CloudWatch.

If I view a trace, for which I previously could view logs, and click on "Resources" for "Cloudwatch logs" I see the following:

```
[
  {
    "log_group": "*",
    "arn": "arn:aws:logs:eu-central-1:<account>:log-group:/aws<the name of my log group>:*"
  }
]
```

I also tried adding another separting `/` between `/aws` and `logGroupName` but that also doesn't work. In CloudWatch I then see:

```
[
  {
    "log_group": "*",
    "arn": "arn:aws:logs:eu-central-1:<account>:log-group:/aws/<the name of my log group>:*"
  }
]
```

But that is also wrong as my log group is just called `<the name of my log group>` and not `/aws/<the name of my log group>`.

**Comment by mmanciop** [comment]
> @mmanciop Thanks for rebasing your changes on my PR!
> 
> Regarding the change in the log group ARN. That dosn't seem right to me. Have you tested this on AWS?
> 
> I just changed this line on my branch, to test this change in isolation, and deployed to our test evironment.
> 
> ```
> // from
> attributesBuilder.put(ResourceAttributes.AWS_LOG_GROUP_ARNS, "arn:aws:logs:" + region + ":" + account + ":log-group:" + logGroupName);
> // to
> attributesBuilder.put(ResourceAttributes.AWS_LOG_GROUP_ARNS, "arn:aws:logs:" + region + ":" + account + ":log-group:/aws" + logGroupName + ":*");
> ```
> 
> With this change in place I can no longer see the logs for a trace in CloudWatch.
> 
> If I view a trace, for which I previously could view logs, and click on "Resources" for "Cloudwatch logs" I see the following:
> 
> ```
> [
>   {
>     "log_group": "*",
>     "arn": "arn:aws:logs:eu-central-1:<account>:log-group:/aws<the name of my log group>:*"
>   }
> ]
> ```
> 
> I also tried adding another separting `/` between `/aws` and `logGroupName` but that also doesn't work. In CloudWatch I then see:
> 
> ```
> [
>   {
>     "log_group": "*",
>     "arn": "arn:aws:logs:eu-central-1:<account>:log-group:/aws/<the name of my log group>:*"
>   }
> ]
> ```
> 
> But that is also wrong as my log group is just called `<the name of my log group>` and not `/aws/<the name of my log group>`.

I compared the ARNs with the output of `aws logs describe-log-groups` and `aws logs describe-log-streams`. At first, when I started investigating the change, I was as surprised as you. @willarmiros do you concur with these ARN values?

**Comment by felixscheinost** [comment]
@mmanciop When I do `describe-log-groups` on my log groups I get an ARN without `/aws`.

**Comment by mmanciop** [comment]
> @mmanciop When I do `describe-log-groups` on my log groups I get an ARN without `/aws`.
> 
> Only when the log group name contains `/aws`, like for lambdas, does the ARN also contain `/aws`.

I tried it on ECS, but this bears more validation. Brb :-)

**Comment by mmanciop** [comment]
So, what I see in the ECS console when deploying a simple CDK + Amazon ECS project is this:

```
awslogs-group | /aws/batch/job
-- | --
awslogs-region | eu-central-1
awslogs-stream-prefix | CrawlerJobDefinition4D5-55fb222b188f18e
```

The log-group is called `/aws/batch/job` in CloudWatch and the ARN is `arn:aws:logs:eu-central-1:170906415945:log-group:/aws/batch/job:*`.

The AWS CLI reports:
```
% aws logs describe-log-groups | jq '.logGroups[] | select(.logGroupName == "/aws/batch/job")'
{
  "logGroupName": "/aws/batch/job",
  "creationTime": 1658054662260,
  "metricFilterCount": 0,
  "arn": "arn:aws:logs:eu-central-1:170906415945:log-group:/aws/batch/job:*",
  "storedBytes": 141110571
}
```

So, yes, we should not add programmatically `/aws`. The `:*` at the end of the ARN for the log-group, however, seem pretty legit. Do you agree @felixscheinost ?

**Comment by mmanciop** [comment]
@felixscheinost I pushed a commit to remove the spurious `/aws` prefix from the ARNs. Thanks for catching that :-)

**Comment by felixscheinost** [comment]
> The :* at the end of the ARN for the log-group, however, seem pretty legit.

Yeah, I think so. It seems like it does work without it but both the CLI and aws.amazon.com display log groups with that trailing `:*`.

Okay, I will manually test the other changes from your commit after the weekend and if everything looks fine (which I assume it will), pull them into this PR.

Is it okay for you if I squash your two commits into one and push it on top of this branch?

**Comment by mmanciop** [comment]
> Is it okay for you if I squash your two commits into one and push it on top of this branch?

Yup, I think in the end it is best to squash the entire PR on merge anyhow

**Comment by felixscheinost** [comment]
Hi @mmanciop, unfortunately even the trailing `:*` seems to break showing the logs for a trace.

I cherry picked all your changes, deployed and tried showing the logs for a trace but didn't see any. As sometimes the logs take a while to show up I waited and tried again but still no logs.

Then I removed the trailing `:*` from the log group ARNs, deployed again and now I can see the logs again.

The output for the trace under "Resources" -> "CloudWatch logs" is:

```
[
  {
    "log_group": "*",
    "arn": "arn:aws:logs:eu-central-1:575990215942:log-group:dev-cdb-log-group:*"
  }
]
```

vs this:

```
[
  {
    "log_group": "dev-cdb-log-group",
    "arn": "arn:aws:logs:eu-central-1:575990215942:log-group:dev-cdb-log-group"
  }
]
```

So it seems CloudWatch is very peculiar about how the log group ARNs are formatted. Maybe someone from AWS can verify this.

**Comment by felixscheinost** [comment]
Hi @mmanciop, unfortunately even the trailing `:*` seems to break showing the logs for a trace.

I cherry picked all your changes, deployed and tried showing the logs for a trace but didn't see any. As sometimes the logs take a while to show up I waited and tried again but still no logs.

Then I removed the trailing `:*` from the log group ARNs, deployed again and now I can see the logs again.

The output for the trace under "Resources" -> "CloudWatch logs" is:

```
[
  {
    "log_group": "*",
    "arn": "arn:aws:logs:eu-central-1:<account number>:log-group:<log group name>:*"
  }
]
```

vs this:

```
[
  {
    "log_group": "<log group name>",
    "arn": "arn:aws:logs:eu-central-1:<account number>:log-group:<log group name>"
  }
]
```

So it seems CloudWatch is very peculiar about how the log group ARNs are formatted. Maybe someone from AWS can verify this.

**Comment by mmanciop** [comment]
> I cherry picked all your changes, deployed and tried showing the logs for a trace but didn't see any. As sometimes the logs take a while to show up I waited and tried again but still no logs.

> Then I removed the trailing :* from the log group ARNs, deployed again and now I can see the logs again.

@felixscheinost how exactly are you using these ARNs? With which tool? Because the issue could be there too.

**Comment by felixscheinost** [comment]
I am using CloudWatch traces on the official AWS web console.

**Comment by felixscheinost** [comment]
> Besides, the spec clearly states with the example that the :* must be there: https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/cloud_provider/aws/logs/

These are merely "examples".

This links also references [the official AWS documentation](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/iam-access-control-overview-cwl.html#CWL_ARN_Format), which states:

> Both of the following are used. The second one, with the * at the end, is what is returned by the describe-log-groups CLI command and the DescribeLogGroups API.
> arn:aws:logs:region:account-id:log-group:log_group_name
> arn:aws:logs:region:account-id:log-group:log_group_name:*

I think in the end what works is what's correct. 3rd party tools which parse `aws.log.group.arns` can be adapted. AWS CloudWatch can hardly be adapted.

**Comment by mmanciop** [comment]
I suppose @willarmiros, @Aneurysm9 should be making the call and, if `:*` has to go, update the examples in the spec.

**Comment by willarmiros** [comment]
Sorry for being late here: so from my reading there were 2 issues:
1. `/aws` was being pre-pended to log groups in the resource detector, but that was fixed to no longer do that (which is the correct fix)
2. There is a question of whether to include the trailing `:*` in the log group ARN when recording it on the trace. It seems like the expected behavior is "it doesn't matter" because both such ARN formats are valid, however in practice only the ARN without the trailing `:*` works. I have just found the bug on our end here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8a821bb7829a23924edd2e23b22c6751277ab2ea/exporter/awsxrayexporter/internal/translator/aws.go#L272-L279

I will submit a fix for the bug so that both ARN formats work. The spec should stay as-is because the examples are still valid formats, just our parsing code is wrong.

File: sdk-extensions/aws/src/test/java/io/opentelemetry/sdk/extension/aws/resource/EcsResourceTest.java [hunk 2]
```diff
@@ -28,45 +31,97 @@ class EcsResourceTest {
...
+    Resource resource = EcsResource.buildResource(mockSysEnv, mockHttpClient);
+    Attributes attributes = resource.getAttributes();
+
+    assertThat(resource.getSchemaUrl()).isEqualTo(ResourceAttributes.SCHEMA_URL);
     assertThat(attributes)
         .containsOnly(
             entry(ResourceAttributes.CLOUD_PROVIDER, "aws"),
             entry(ResourceAttributes.CLOUD_PLATFORM, "aws_ecs"),
-            entry(ResourceAttributes.CONTAINER_NAME, InetAddress.getLocalHost().getHostName()));
+            entry(ResourceAttributes.CONTAINER_NAME, "ecs-curltest-24-curl-cca48e8dcadd97805600"),
```
Comment (jj22ee) [inline sdk-extensions/aws/src/test/java/io/opentelemetry/sdk/extension/aws/resource/EcsResourceTest.java:88]: Change "24" to "26" for consistency.

File: sdk-extensions/aws/src/test/resources/ecs-container-metadata-v4.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
+{
+  "DockerId": "ea32192c8553fbff06c9340478a2ff089b2bb5646fb718b4ee206641c9086d66",
+  "Name": "curl",
+  "DockerName": "ecs-curltest-24-curl-cca48e8dcadd97805600",
```
Comment (jj22ee) [inline sdk-extensions/aws/src/test/resources/ecs-container-metadata-v4.json:4]: Change "24" to "26" for consistency.

File: sdk-extensions/aws/src/test/resources/ecs-container-metadata-v4.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  "Image": "111122223333.dkr.ecr.us-west-2.amazonaws.com/curltest:latest",
+  "ImageID": "sha256:d691691e9652791a60114e67b365688d20d19940dde7c4736ea30e660d8d3553",
+  "Labels": {
+    "com.amazonaws.ecs.cluster": "default",
+    "com.amazonaws.ecs.container-name": "curl",
+    "com.amazonaws.ecs.task-arn": "arn:aws:ecs:us-west-2:111122223333:task/default/8f03e41243824aea923aca126495f665",
+    "com.amazonaws.ecs.task-definition-family": "curltest",
+    "com.amazonaws.ecs.task-definition-version": "24"
```
Comment (jj22ee) [inline sdk-extensions/aws/src/test/resources/ecs-container-metadata-v4.json:12]: Should be "26" to match ```ecs-task-metadata-v4.json```

**Review by jj22ee — APPROVED** [review]
Looks good to me!
Just requested a small change to keep the task-definition-version number consistent in the V4 files.

**Comment by willarmiros** [comment]
Hi @mmanciop - I have opened the PR to address the `:*` issue on the X-Ray exporter side, and @jj22ee has left a few comments but approved. Would you mind addressing the comments so we can then merge?

**Comment by mmanciop** [comment]
@willarmiros I cannot push to @felixscheinost's branch (I have to push rights for his fork, which he used to open this PR), but I prepared the commit on https://github.com/lumigo-io/opentelemetry-java/tree/pr-for-felix .

**Comment by felixscheinost** [comment]
I addressed @jj22ee's comments and cherry-picked @mmanciop's additions.

**Comment by felixscheinost** [comment]
@willarmiros 

>  However, for now it would be best if the resource detectors could record the ARNs without the trailing :* so that everything still works while we wait for the fix to land.

I pushed another commit which removes the trailing `:*` for now.

**Comment by jkwatson** [comment]
@willarmiros when you think this is ready, please approve and we can move forward. Thanks!

**Review by willarmiros — APPROVED** [review]

**Review by jkwatson — APPROVED** [review]
Thanks!

**Pull request merged**